### DISCUSSION TITLE
실행 컨텍스트가 생성될 때 자바스크립트 엔진이 수행하는 주요 작업은 무엇인가요?

### DISCUSSION THREAD

**Discussion by lledellebell** [diving-javascript/you-dont-know-js-yet#22] [실행 컨텍스트(Execution Context)]
- #18

**Comment by wogus216** [comment]
자바스크립트 엔진은 실행 컨텍스트를 다룰 때 여러 단계로 나누어 작업을 진행합니다. 이러한 단계적 접근은 코드를 효율적으로 처리하고, 각 코드 블록이 필요로 하는 환경을 정확하게 설정하기 위해 필수적입니다.

생성 단계 (Creation Phase): 이 단계에서 자바스크립트 엔진은 스코프, 변수, 함수 선언 등이 담긴 변수 객체를 생성하고, 스코프 체인을 형성하며, this 키워드의 값을 결정합니다. 이는 코드 실행 전에 실행 컨텍스트를 준비하는 초기 설정 단계라고 할 수 있습니다.
활성화 단계 (Activation / Code Execution Phase): 실제 코드가 실행되는 단계입니다. 변수 인스턴스화와 함수 선언의 메모리 할당이 완료된 후, 코드 내의 각종 연산이나 함수 호출 등이 실행됩니다. 이 단계에서 변수에 값이 할당되고, 실제 연산이 처리됩니다.
소멸 단계 (Teardown Phase): 함수 호출이 완료되고 해당 실행 컨텍스트가 필요 없어지면, 자바스크립트 엔진은 해당 컨텍스트를 스택에서 제거하고 관련 자원을 정리합니다. 이는 메모리를 효율적으로 관리하고, 불필요한 자원 사용을 줄이기 위한 단계입니다.
